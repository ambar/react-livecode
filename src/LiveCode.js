/** @jsx jsx */
import {jsx} from '@emotion/core'
import React, {useState} from 'react'
import prettyFormat from 'pretty-format'
import dedent from 'dedent'
import isPromise from 'is-promise'
import useLiveCode from './useLiveCode'
import LiveCodeEditor from './LiveCodeEditor'

const inspectPromise = {
  test: isPromise,
  serialize(array, config, indentation, depth, refs, printer) {
    // TODO: inspect async value
    return 'Promise'
  },
}

const safeTry = (fn, fallback) => {
  try {
    return fn()
  } catch (e) {
    return fallback(e)
  }
}

const prettyPrint = obj =>
  prettyFormat(obj, {
    min: true,
    printFunctionName: true,
    plugins: [inspectPromise],
  })

const LiveCode = ({
  code: initialCode = '',
  scope = {},
  split = 'horizontal',
  autoFocus = false,
  modules = true,
  ...nativeProps
}) => {
  const [code, setCode] = useState(() => dedent(initialCode))
  const isVertical = split === 'vertical'

  const serialize = ret => {
    // module script loading
    if (ret == null) {
      return ' '
    }

    // evaluation ok
    if (Array.isArray(ret)) {
      return ret.length
        ? ret.map((line, i) => <div key={i}>&rsaquo; {prettyPrint(line)}</div>)
        : ' '
    }

    // evaluation error
    return <div>&rsaquo; {prettyPrint(ret)}</div>
  }

  const values = useLiveCode(code, scope, {modules})

  return (
    <div
      css={{
        display: 'flex',
        justifyContent: 'center',
        margin: '1em auto',
        maxWidth: '100vw',
        maxHeight: '100vh',
        fontSize: '0.85em',
        lineHeight: '1.45',
        borderRadius: '3px',
        boxShadow: '0 1px 5px rgba(0,0,0,.1)',
        fontFamily: '"Fira code", "Fira Mono", Menlo, Consolas, monospace',
        flexDirection: isVertical ? 'column' : 'row',
      }}
      {...nativeProps}
    >
      <LiveCodeEditor
        autoFocus={autoFocus}
        value={code || ' '}
        onValueChange={code => setCode(code)}
        padding={10}
        css={{
          width: isVertical ? '100%' : '50%',
          minHeight: '5em',
          color: '#222',
          backgroundColor: '#fafafa',
          outline: 'none',
          fontSize: 'inherit !important',
          overflow: 'auto',
          borderRadius: isVertical ? '0 3px 3px 0' : '3px 0 0 3px',
          textarea: {
            outline: 'none',
          },
          '.token': {
            backgroundColor: 'inherit!important',
          },
        }}
      />
      <pre
        css={{
          width: isVertical ? '100%' : '50%',
          margin: '0',
          padding: '10px',
          fontFamily: 'inherit',
          fontWeight: 300,
          color: '#eee',
          backgroundColor: '#282c34',
          overflow: 'auto',
          WebkitOverflowScrolling: 'auto',
          borderRadius: isVertical ? '0 0 3px 3px' : '0 3px 3px 0',
          boxSizing: 'border-box',
        }}
      >
        {safeTry(() => serialize(values), serialize)}
      </pre>
    </div>
  )
}

export default LiveCode
