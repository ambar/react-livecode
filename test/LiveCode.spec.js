import React from 'react'
import {render} from 'react-testing-library'
import chunk from 'lodash/chunk'
import LiveCode from '../src/LiveCode'

const sampleCode = `
  2 ** 3
  chunk(['a', 'b', 'c', 'd'], 2)
  new Map([['key1', 'value1'], ['key2', 'value2']])
  const never = new Promise(_ => {})
  never
`

describe('transformCode', () => {
  it('should render synchronously', () => {
    const {container} = render(
      <LiveCode autoFocus scope={{chunk}} code={sampleCode} modules={false} />
    )
    expect(
      container.querySelector('pre[class*=LiveCode]').children
    ).toMatchSnapshot()
  })
})
