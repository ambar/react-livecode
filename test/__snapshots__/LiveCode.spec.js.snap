// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformCode should render synchronously 1`] = `
HTMLCollection [
  <div>
    › 
    8
  </div>,
  <div>
    › 
    [["a", "b"], ["c", "d"]]
  </div>,
  <div>
    › 
    Map {"key1" =&gt; "value1", "key2" =&gt; "value2"}
  </div>,
  <div>
    › 
    Promise
  </div>,
]
`;
